internal/controller/apiextensions/composite/environment/selector.go

```diff
@@ -114,15 +114,18 @@ func (s *APIEnvironmentSelector) buildEnvironmentConfigRefFromRef(ref *v1.Enviro
 }
 
 func (s *APIEnvironmentSelector) lookUpConfigs(ctx context.Context, cr resource.Composite, ml []v1.EnvironmentSourceSelectorLabelMatcher) (*v1alpha1.EnvironmentConfigList, error) {
+	res := &v1alpha1.EnvironmentConfigList{}
 	matchLabels := make(client.MatchingLabels, len(ml))
 	for i, m := range ml {
 		val, err := ResolveLabelValue(m, cr)
 		if err != nil {
+			if fieldpath.IsNotFound(err) && m.FromFieldPathIsOptional() {
+				return res, nil
```

Comment:
Shouldn't we `continue` here instead of returning res which effectively has an empty environment config list?

Reply:
I had the same doubt, I decided to make it "no-op" the whole environment, as I thought that omitting a single value would have been more confusing. This was my thought:
- if a single label selector is specified and marked as optional, we should return no environment at all.
- if `n>1` selectors are specified, one of them is marked as optional, if not found I'd rather have it result in a "no-op" than selecting an environment selected only through the `n-1` remaining selectors.
But I agree it's debatable, what would you suggest?

Reply:
> if n>1 selectors are specified, one of them is marked as optional, if not found I'd rather have it result in a "no-op" than selecting an environment selected only through the n-1 remaining selectors.

Since the policy is defined inside an element of `matchLabels` array, I would expect it only to configure the behavior for that single element. Take the following example:

```yaml
          matchLabels:
            - type: Value
              key: cloud-provider
              value: aws
            - type: FromCompositeFieldPath
              key: stage
              valueFromFieldPath: spec.someValue
              fromFieldPathPolicy: Optional
```

I would expect this still to filter `cloud-provider=aws` if `spec.someValue` is absent.

> if a single label selector is specified and marked as optional, we should return no environment at all

This definitely makes sense, instead of selecting all by considering this as "no selector provided".

However, I see the inconsistency between the two cases, if a single element - we filter all out, if multiple elements, we filter all in from the entry with policy POV. 
Also it is not only 1 vs n selectors being provided, e.g. what happens if multiple - all optional / none resolved:

```yaml
          matchLabels:
            - type: FromCompositeFieldPath
              key: cloud-provider
              valueFromFieldPath: spec.anotherValue
              fromFieldPathPolicy: Optional
            - type: FromCompositeFieldPath
              key: stage
              valueFromFieldPath: spec.someValue
              fromFieldPathPolicy: Optional
```

So, I agree that there is no perfect answer, but I still lean on policy impacting that single entry instead of invalidating all others. Otherwise, we may consider moving the policy up instead to make it more explicit:

```yaml
    environmentConfigs:
      - ref:
          name: example-environment-1
      - type: Selector
        selector:
          compositeValueResolutionPolicy: Optional # This environment config selector will be ignored if composite value resolution fails for any label
          matchLabels:
            - type: FromCompositeFieldPath
              key: stage
              valueFromFieldPath: metadata.labels[stage]
```